app/gui/src/project-view/components/visualizations/TableVisualization/tableVizToolbar.ts

```diff
@@ -79,240 +82,193 @@ function useSortFilterNodesButton({
-          patterns.push(filterPatterns)
-        }
+    const patterns: Pattern[] = []
+    const filters = toValue(filterModel)
+    const sorts = toValue(sortModel)
 
-        const value = getFilterValue(filterModel)
-
-        if (value) {
-          const filterPatterns =
-            sortModelValue.length ?
-              getAstPatternFilterAndSort(columnName, value, filterType, filterAction)
-            : getAstPatternFilter(columnName, value, filterType, filterAction)
-          patterns.push(filterPatterns)
-        }
-      })
-    } else if (sortModelValue.length) {
+    if (sorts.length > 0) {
       patterns.push(getAstPatternSort())
     }
 
-    function getRemoveColumnsAstPattern() {
-      return Pattern.new<Ast.Expression>((ast) => {
-        const columns = Ast.Vector.build(columnsToRemove, Ast.TextLiteral.new, ast.module)
-        return Ast.App.positional(
-          Ast.PropertyAccess.new(ast.module, ast, Ast.identifier('remove_columns')!),
-          columns,
-        )
-      })
-    }
-
-    function getColumnOrderAstPattern() {
-      return Pattern.new<Ast.Expression>((ast) => {
-        const columns = Ast.Vector.build(columnOrder!, Ast.TextLiteral.new, ast.module)
-        return Ast.App.positional(
-          Ast.PropertyAccess.new(ast.module, ast, Ast.identifier('reorder_columns')!),
-          columns,
-        )
-      })
+    const columnsToRemove = toValue(hiddenColumns)
+    if (columnsToRemove.length) {
+      patterns.push(
+        Pattern.new((ast) =>
+          createSimpleAstCall(
+            'remove_columns',
+            Ast.Vector.build(columnsToRemove, Ast.TextLiteral.new, ast.module),
+            ast,
+          ),
+        ),
+      )
     }
 
-    const columnsToRemove = toValue(hiddenColumns)
     const columnOrder = toValue(vizColumnOrder)
+    if (columnOrder?.length) {
+      patterns.push(
+        Pattern.new((ast) =>
+          createSimpleAstCall(
+            'reorder_columns',
+            Ast.Vector.build(columnOrder, Ast.TextLiteral.new, ast.module),
+            ast,
+          ),
+        ),
+      )
+    }
 
-    if (columnsToRemove.length) {
-      patterns.push(getRemoveColumnsAstPattern())
+    function projector(parentPattern: ConstructivePattern | undefined) {
+      return (columnName: any, filterAction: any, filterType: any, value: any) =>
+        (ast: Ast.Owned<Ast.MutableExpression>) => {
+          const parentPat = parentPattern ? parentPattern(ast) : ast
+          return buildPattern(ast, parentPat, columnName, filterAction, filterType, value)
+        }
     }
-    if (columnOrder != null) {
-      patterns.push(getColumnOrderAstPattern())
+
+    let filterPatterns = new Array<ConstructivePattern>()
+    for (const filter of filters) {
+      const { columnName, filterAction, filterType } = filter
+      const value = getFilterValue(filter)
+
+      filterPatterns = (filterPatterns.length ? filterPatterns : [undefined]).flatMap((parent) => {
```

Comment:
It looks like the mapping function always returns exactly one element--could this use `map`?